charmcraft/charm_builder.py

```diff
@@ -239,26 +239,26 @@ def _install_dependencies(self, staging_venv_dir):
         with instrum.Timer("Installing all dependencies"):
             if self.binary_python_packages:
                 # install python packages, allowing binary packages
-                cmd = [pip_cmd, "install", "--upgrade"]  # base command
+                cmd = [pip_cmd, "install"]  # base command
                 cmd.extend(self.binary_python_packages)  # the python packages to install
                 _process_run(cmd)
 
             if self.python_packages:
                 # install python packages from source
-                cmd = [pip_cmd, "install", "--upgrade", "--no-binary", ":all:"]  # base command
+                cmd = [pip_cmd, "install", "--no-binary", ":all:"]  # base command
```

Comment:
It looks like the Jupyter notebooks charm is failing because of this (see [spread test](https://github.com/canonical/charmcraft/actions/runs/5468408751/jobs/9956182984?pr=1136#step:5:518).

Between this and the above, I'm wondering whether the right change for now might be to keep the `--upgrade` here and above and merge the two commands below (given the edge case we discussed in the main conversation).

I believe that'll resolve #1135 with a fairly minimal change that we can put into a 2.4 release fairly quickly, with a bigger rethink of package handling for a future release. How's that sound?

Reply:
Or we can change the order of install `requirement.txt` and `PYDEPS`.

Reply:
So if a pinned version is installed, `pip install` without `--upgrade` will not try to get a new version. And if the package is not in `requirement.txt`, it will install latest stable version.

Reply:
Or maybe we can add `PYDEPS` to a new `requirement.txt`